### ISSUE TITLE
use update_branch_immediately=true with require_automerge_label=true

### ISSUE THREAD

**Issue by clxmstaab** [chdsbd/kodiak#174]
we use the following kodiak config

```
# .kodiak.toml
version = 1

[merge]
method = "squash"
# once a PR is merged into master, delete the branch
delete_branch_on_merge = true
# label to use to enable Kodiak to merge a PR
automerge_label = "automerge" # default: "automerge"
# require that the automerge label be set for Kodiak to merge a PR. if you
# disable this Kodiak will immediately attempt to merge every PR you create
require_automerge_label = true
# if a PR is passing all checks and is able to be merged, merge it without
# placing it in the queue. This will introduce some unfairness where those
# waiting in the queue the longest will not be served first.
prioritize_ready_to_merge = true
# immediately update a PR whenever the target updates. If enabled, Kodiak will
# not be able to efficiently update PRs. Any time the target of a PR updates,
# the PR will update.
update_branch_immediately = true

[merge.message]
title = "pull_request_title"
body = "github_default"
include_pr_number = true
strip_html_comments = true # remove html comments to auto remove PR templates
```

and we have the following use-case:
we would like kodiak to automatically merge changes from the target branch into the PR, but do not yet merge them. therefore we use a `WIP:` title-prefix and mark PRs as soon as we want kodiak to update them with the merge-label.

it seems kodiak does not start auto-merging in changes from the target branch as long the title uses `WIP:` prefix... any chance we can get the desired behaviour?

**Comment by sbdchd** [comment]
Hmm, ‘update_branch_immediately’ should update the PR when the target branch changes.

My first thought is the title blacklist regex is taking precedence over the ‘update_branch_immediately’ feature.

Sounds like a bug in Kodiak

**Comment by chdsbd** [comment]
Okay, I've looked into #175 and I think we should actually add another configuration option for this behavior. Right now `update_branch_immediately` affects PRs that are mergeable and need an update. I think the behavior you want is to update a PR even if it is marked as "not mergeable" by having WIP in the title.

Would you want kodiak to update a branch even if it has failing status checks, or should this configuration option be limited to WIP regex?

I was thinking we could add a single flag to cover the case of the WIP regex, but we may want to make it a list of options too.

```toml
update_branch_immediately_ignore_wip_regex = true
# or
update_branch_immediately_ignore_blocker = ["wip_regex", "status_checks"]
```

I'm curious to hear your thoughts.

**Comment by clxmstaab** [comment]
for my use-case it would be great to update the branch even if the build is failing.

in our TRAVIS CI setup a CI running build will automatically be cancelled when new commits are added to a PR. that way we don't get a overlong queue but built only the latest stuff in the Pull requests.
In case we want kodiak to wait for all builds to succeed we would get a really long wait time I guess.

we only need to make sure that kodiak does not merge later on as long as the build fails (which kodiak already does)

**Comment by clxmstaab** [comment]
but hey: most of our PRs are green before kodiak gets activated via Pull Request label - though maybe it will just work

**Comment by clxmstaab** [comment]
any progress/ideas on how to proceed?

**Comment by chdsbd** [comment]
I looked into implementing this and it was more work than expected. My current plan is to add a `update_branch_immediately_ignore_blocker` option, but it will take a bit. I’ll take a look again this evening.

**Comment by clxmstaab** [comment]
Hey Christopher,

I guess for my use-case https://github.com/chdsbd/kodiak/pull/175 will already be enough.
in case you dont have time to work on this other options you are refering to, merging my PR would be a great step forward for us.

thanks for considering

**Comment by chdsbd** [comment]
I haven't looked at this since September, but I think #175 won't resolve your issue. I believe changes need to be made outside of the `mergeable` function. Maybe in [`process_queue`](https://github.com/chdsbd/kodiak/blob/eeda75f13e32e74733a4c3fa612d95ac094549c4/kodiak/queue.py#L178) or [`process_webhook_event`](https://github.com/chdsbd/kodiak/blob/eeda75f13e32e74733a4c3fa612d95ac094549c4/kodiak/queue.py#L73).

I've been working on other things in my free time so I can't commit to implementing this feature.

**Comment by staabm** [comment]
@chdsbd Any chance you could implement it when we pay you for it?

**Comment by chdsbd** [comment]
@staabm That could work. Not sure how to arrange that, but I think that would be a way forward.

**Comment by staabm** [comment]
Could you add the project to github sponsors?

**Comment by chdsbd** [comment]
@staabm Will do. I've just requested approval for GitHub sponsors and they say it will take 7-10 days for the profile to be approved. I'll ping this issue when I get an update.

**Comment by chdsbd** [comment]
I've been approved and have setup GitHub sponsors for Kodiak.

**Comment by staabm** [comment]
Just got the GO from my boss. Will configure the sponsoring tomorrow

**Comment by staabm** [comment]
we just started the sponsoring for the project (complexgmbh).

it would be great if this small financial support would help to get this feature implemented in the near future.

thanks for you great work on this great project.

**Comment by chdsbd** [comment]
Thanks for the support!

I’ll take a look this weekend.

**Comment by MaicolBen** [comment]
Aside from this, it would be great to block some PRs from auto-updating

**Comment by chdsbd** [comment]
@staabm I'm finally looking to implement this feature in #198 now that #207 is merged.

Would new flags like the following work?

```toml
version = 1

[update]
# update PR whenever the PR is out of date with the base branch. PR will be
# updated regardless of failing status checks, missing reviews, merge blacklist
# labels, etc. Kodiak will only update the PR if the automerge label is attached
# or `update.require_automerge_label` is false.
always = false
require_automerge_label = true
```

I think more settings could be added like the following, but they would take more time to implement, so if the flags described above work for you, let's go with them for now.

```toml
[update]
# ...
ignore_blacklist_title_regex = false
ignore_failing_checks = []
ignore_missing_automerge_label = true
ignore_blacklist_labels = true
```

**Comment by staabm** [comment]
> Would new flags like the following work?
> 
> ```toml
> version = 1
> 
> [update]
> # update PR whenever the PR is out of date with the base branch. PR will be
> # updated regardless of failing status checks, missing reviews, merge blacklist
> # labels, etc. Kodiak will only update the PR if the automerge label is attached
> # or `update.require_automerge_label` is false.
> always = false
> require_automerge_label = true
> ```

I think this will fit our use-case

**Comment by chdsbd** [comment]
#198 has been released so you should now be able to use `update.always` and `update.require_automerge_label` as described above.

**Comment by chdsbd** [comment]
@MaicolBen I think you might be able to remove the automerge_label if you need to block a PR from updating. If that doesn't fit with your workflow, please feel free to open a new issue.

**Comment by clxmstaab** [comment]
> [update]
# update PR whenever the PR is out of date with the base branch. PR will be
# updated regardless of failing status checks, missing reviews, merge blacklist
# labels, etc. Kodiak will only update the PR if the automerge label is attached
# or `update.require_automerge_label` is false.
always = false
require_automerge_label = true

just to make sure I got it right.. I guess the example you pointed out is not correct and the opposite is what we need, right?

->
```
always = true
require_automerge_label = false
```

**Comment by chdsbd** [comment]
Thanks. Your example is correct. I’ll edit the other message to be accurate.

**Comment by clxmstaab** [comment]
we are using the new config for a few days and observe the following issues

kodiak does only auto-update a PR/branch when e.g. new commits are added to the PR or the PR description/title changed etc.

kodiak does not update a PR when a new commit lands in the target-branch.

Is this the expected behaviour?

**Comment by chdsbd** [comment]
That’s not intended. Can you post your Kodiak config?

My suspicion is that when the target branch updates, GitHub isn’t sending a webhook event for the open PRs. I probably need to configure GitHub to send webhook events for pushes.

I’ll take a look at this tonight.

**Comment by staabm** [comment]
our latest config:
```
# .kodiak.toml
version = 1

[merge]
method = "squash"
# once a PR is merged into master, delete the branch
delete_branch_on_merge = true
# label to use to enable Kodiak to merge a PR
automerge_label = "automerge" # default: "automerge"
# require that the automerge label be set for Kodiak to merge a PR. if you
# disable this Kodiak will immediately attempt to merge every PR you create
require_automerge_label = true
# if a PR is passing all checks and is able to be merged, merge it without
# placing it in the queue. This will introduce some unfairness where those
# waiting in the queue the longest will not be served first.
prioritize_ready_to_merge = true

[merge.message]
title = "pull_request_title"
body = "github_default"
include_pr_number = true
strip_html_comments = true # remove html comments to auto remove PR templates

[update]
# update PR whenever the PR is out of date with the base branch. PR will be
# updated regardless of failing status checks, missing reviews, merge blacklist
# labels, etc. Kodiak will only update the PR if the automerge label is attached
# or `update.require_automerge_label` is false.
always = true
require_automerge_label = true
``

**Comment by rgreinho** [comment]
We are seeing the same behavior.

Here is our configuration file:

```toml
version = 1

[merge]
require_automerge_label = false
blacklist_title_regex = "^WIP:.*"
blacklist_labels = ["do-not-merge"]
method = "squash"
delete_branch_on_merge = true
prioritize_ready_to_merge = true

[update]
always = true
require_automerge_label = false
```

**Comment by chdsbd** [comment]
I've opened a new issue to track this: https://github.com/chdsbd/kodiak/issues/244
